[quant][graphmode][fx] Add default_reuse_input_qconfig for ops like reshape

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69249

Summary:
This PR added default_reuse_input_qconfig and default_reuse_input_observer which is used
when we want to configure an operator to reuse the observer from input, if the input
Tensor for the operator is not observed, we will not observe the output of this operator either,
if the input Tensor is observed, we will observe the output of the operator with the same observer.

e.g.

```
x1 = x0.reshape()
```
if reshape is configured with default_reuse_input_qconfig:
1. if x0 is observed with observer_0, we'll observe x1 with the same observer instance
2. if x0 is not observed, we won't observe x1 either

Test Plan:
```
python test/test_quantization.py TestQuantizeFx.test_reuse_input_qconfig
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32774723](https://our.internmc.facebook.com/intern/diff/D32774723)

jerryzh168 committed:
[quant][graphmode][fx] Add default_replay_qconfig for ops like reshape

Summary:
This PR added default_replay_qconfig and default_replay_observer which is used
when we want to configure an operator to reuse the observer from input, if the input
Tensor for the operator is not observed, we will not observe the output of this operator either,
if the input Tensor is observed, we will observe the output of the operator with the same observer.

e.g.

```
x1 = x0.reshape()
```
if reshape is configured with default_replay_qconfig:
1. if x0 is observed with observer_0, we'll observe x1 with the same observer instance
2. if x0 is not observed, we won't observe x1 either

Test Plan:
```
python test/test_quantization.py TestQuantizeFx.test_replay_qconfig
```
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32774723).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add default_replay_qconfig for ops like reshape"


Summary:
This PR added default_replay_qconfig and default_replay_observer which is used
when we want to configure an operator to reuse the observer from input, if the input
Tensor for the operator is not observed, we will not observe the output of this operator either,
if the input Tensor is observed, we will observe the output of the operator with the same observer.

e.g.

```
x1 = x0.reshape()
```
if reshape is configured with default_replay_qconfig:
1. if x0 is observed with observer_0, we'll observe x1 with the same observer instance
2. if x0 is not observed, we won't observe x1 either

Test Plan:
```
python test/test_quantization.py TestQuantizeFx.test_replay_qconfig
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32774723](https://our.internmc.facebook.com/intern/diff/D32774723)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32774723).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add default_reuse_input_qconfig for ops like reshape"


Summary:
This PR added default_reuse_input_qconfig and default_reuse_input_observer which is used
when we want to configure an operator to reuse the observer from input, if the input
Tensor for the operator is not observed, we will not observe the output of this operator either,
if the input Tensor is observed, we will observe the output of the operator with the same observer.

e.g.

```
x1 = x0.reshape()
```
if reshape is configured with default_reuse_input_qconfig:
1. if x0 is observed with observer_0, we'll observe x1 with the same observer instance
2. if x0 is not observed, we won't observe x1 either

Test Plan:
```
python test/test_quantization.py TestQuantizeFx.test_reuse_input_qconfig
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32774723](https://our.internmc.facebook.com/intern/diff/D32774723)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32774723).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add default_reuse_input_qconfig for ops like reshape"


Summary:
This PR added default_reuse_input_qconfig and default_reuse_input_observer which is used
when we want to configure an operator to reuse the observer from input, if the input
Tensor for the operator is not observed, we will not observe the output of this operator either,
if the input Tensor is observed, we will observe the output of the operator with the same observer.

e.g.

```
x1 = x0.reshape()
```
if reshape is configured with default_reuse_input_qconfig:
1. if x0 is observed with observer_0, we'll observe x1 with the same observer instance
2. if x0 is not observed, we won't observe x1 either

Test Plan:
```
python test/test_quantization.py TestQuantizeFx.test_reuse_input_qconfig
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32774723](https://our.internmc.facebook.com/intern/diff/D32774723)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32774723).

vkuzo commented on torch/ao/quantization/fx/qconfig_utils.py:
```diff
@@ -237,9 +237,12 @@ def generate_qconfig_map(
 
         elif node.op == "call_method":
             module_path, module_type = node_name_to_scope[node.name]
+            # first use node.target (string) to get the qconfig
+            qconfig = maybe_adjust_qconfig_for_module_type_or_name(
+                qconfig_dict, node.target, module_path, global_qconfig)
```
is this related to this PR?  If yes, any chance we can make the comment more descriptive?

jerryzh168 replied:
yeah this is to enable that we can configure "object_type" with call_method strings, this functionality is missing previously and looks like no one has used this before

vkuzo commented on torch/ao/quantization/qconfig.py:
```diff
@@ -170,6 +172,12 @@ def __new__(cls, activation=torch.nn.Identity, weight=torch.nn.Identity):
 """
 
+default_reuse_input_qconfig = QConfig(activation=default_reuse_input_observer,
+                                      weight=default_reuse_input_observer)
```
would it make sense to set weight to `None`?  Not sure if there are cases where the right thing to do for current op is to reuse the weight observer of the previous op.

jerryzh168 replied:
we don't have None observers right now I think, maybe I can set it to a NoopObserver

vkuzo reviewed (approved):
thanks for working on this!

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add default_reuse_input_qconfig for ops like reshape"


Summary:
This PR added default_reuse_input_qconfig and default_reuse_input_observer which is used
when we want to configure an operator to reuse the observer from input, if the input
Tensor for the operator is not observed, we will not observe the output of this operator either,
if the input Tensor is observed, we will observe the output of the operator with the same observer.

e.g.

```
x1 = x0.reshape()
```
if reshape is configured with default_reuse_input_qconfig:
1. if x0 is observed with observer_0, we'll observe x1 with the same observer instance
2. if x0 is not observed, we won't observe x1 either

Test Plan:
```
python test/test_quantization.py TestQuantizeFx.test_reuse_input_qconfig
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32774723](https://our.internmc.facebook.com/intern/diff/D32774723)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32774723).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add default_reuse_input_qconfig for ops like reshape"


Summary:
This PR added default_reuse_input_qconfig and default_reuse_input_observer which is used
when we want to configure an operator to reuse the observer from input, if the input
Tensor for the operator is not observed, we will not observe the output of this operator either,
if the input Tensor is observed, we will observe the output of the operator with the same observer.

e.g.

```
x1 = x0.reshape()
```
if reshape is configured with default_reuse_input_qconfig:
1. if x0 is observed with observer_0, we'll observe x1 with the same observer instance
2. if x0 is not observed, we won't observe x1 either

Test Plan:
```
python test/test_quantization.py TestQuantizeFx.test_reuse_input_qconfig
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32774723](https://our.internmc.facebook.com/intern/diff/D32774723)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32774723).
